custom_components/zwave_mqtt/climate.py

```diff
@@ -0,0 +1,461 @@
+            climate = ZWaveClimateSingleSetpoint(values)
+        elif values.primary.command_class == CommandClass.THERMOSTAT_MODE:
+            climate = ZWaveClimateMultipleSetpoint(values)
+
+        if climate is not None:
+            async_add_entities([climate])
+
+    hass.data[DOMAIN][config_entry.entry_id][DATA_UNSUBSCRIBE].append(
+        async_dispatcher_connect(hass, "zwave_new_climate", async_add_climate)
+    )
+    await hass.data[DOMAIN][config_entry.entry_id]["mark_platform_loaded"]("climate")
+
+
+class ZWaveClimateBase(ZWaveDeviceEntity, ClimateDevice):
+    """Representation of a Z-Wave Climate device."""
+
+    def __init__(self, values):
+        """Initialize the Z-Wave climate device."""
+        super().__init__(values)
+        self._target_temperature = None
+        self._target_temperature_range = (None, None)
+        self._current_temperature = None
+        self._hvac_action = None
+        self._hvac_list = None
+        self._hvac_mode = None
+        self._default_hvac_mode = None
+        self._preset_list = None
+        self._preset_mode = None
+        self._current_fan_mode = None
+        self._fan_modes = None
+        self._fan_action = None
+        self._unit = None
+        self.update_properties()
+
+    @callback
+    def value_changed(self, value):
+        """Update after value change."""
+        self.update_properties()
+        super().value_changed(value)
+
+    def _mode(self) -> None:
+        """Return thermostat mode Z-Wave value."""
+        raise NotImplementedError()
+
+    def _current_mode_setpoints(self) -> Tuple:
+        """Return a tuple of current setpoint Z-Wave value(s)."""
+        raise NotImplementedError()
+
+    def update_properties(self):
+        """Handle the data changes for node values."""
+        self._update_operation_mode()
+        self._update_current_temp()
+        self._update_fan_mode()
+        self._update_target_temp()
+        self._update_operating_state()
+        self._update_fan_state()
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        support = SUPPORT_TARGET_TEMPERATURE
+        if self._hvac_list and HVAC_MODE_HEAT_COOL in self._hvac_list:
+            support |= SUPPORT_TARGET_TEMPERATURE_RANGE
+        if self._preset_list and PRESET_AWAY in self._preset_list:
+            support |= SUPPORT_TARGET_TEMPERATURE_RANGE
+        if self.values.fan_mode:
+            support |= SUPPORT_FAN_MODE
+        if self._preset_list:
+            support |= SUPPORT_PRESET_MODE
+        return support
+
+    def _update_operation_mode(self):
+        """Update hvac and preset modes."""
+        if not self._mode():
+            return
+        self._hvac_list = []
+        self._preset_list = []
+        self._hvac_value_label_mapping = {}
+        self._hvac_label_value_mapping = {}
+        values_list = self._mode().value.get(VALUE_LIST)
```

Comment:
Do we know that `VALUE_LIST` always will be in the mode value? If so we should access it with `self._mode().value[VALUE_LIST]`.

If not, we need to gaurd from setting `values_list` to `None`. We're expecting it to be a list below.

Reply:
I'm erroring on the side of it always being available. I think it should be assuming proper implementation from OZW. If it were to change, then many parts of this would likely need to updated.